### ISSUE TITLE
Enable authentication via environment variables

### ISSUE THREAD

**Issue by garymoon** [Azure/azure-cli#10241]
**Is your feature request related to a problem? Please describe.**
I'm concerned that I cannot use the CLI without having my credentials written to disk in the clear.

**Describe the solution you'd like**
I would like the CLI to accept credentials from the environment (e.g. `AZURE_TENANT_ID`, `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`).

**Describe alternatives you've considered**
I've yet to find a way to authenticate that doesn't involve secrets being written to `~/.azure`.

**Additional context**
I came across [this](https://github.com/Azure/azure-cli/issues/9282) issue which seems to describe the same thing, but was closed following an apparent misunderstanding. While the AWS CLI does indeed write secrets in the clear to disk when configured via `aws configure`, it also allows for [configuration via environment variables](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html).

I've also found examples in both the Ruby SDK [here](https://github.com/Azure/azure-sdk-for-ruby#option-1---environment-variables), and the Go SDK [here](https://github.com/Azure/azure-sdk-for-go#more-authentication-details) where authentication via environment variables is enabled.

This issue is particularly problematic in a dockerized environment where it's common for the home directory to be mapped on the host.

**Comment by jiasli** [comment]
Hi @garymoon, thank you for raising this issue. We will consider this as a feature request and discuss internally.

**Comment by garymoon** [comment]
@jiasli Thank you for the update. Please also be sure to consider the implications of unintended credential leakage also. IMO this issue represents a security concern.

**Comment by jiasli** [comment]
**Python SDK - Azure Identity** also supports environment variables [here](https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/identity/azure-identity#environment-variables). Maybe we can implement this independently of Azure Identity integration. Even after the Azure Identity integration, this won't be a breaking change.

**Comment by kostyay** [comment]
Any update on this? It will be super useful for us as well.

**Comment by jiasli** [comment]
We are currently implementing this feature in the Azure Identity Integration process and it will be included in the next Beta release. But we can't provide a solid ETA yet.

**Comment by etaham** [comment]
Any update?

**Comment by jiasli** [comment]
This will be released in the next Beta version which uses MSAL. Hopefully it can satisfy your needs.

**Comment by aliensouls** [comment]
Year 2021  Q2

lol is this merged already or still discussed internally? 😆   I want to work for Microsoft, please hire me! I want to chill all day like you guys do :)))))))  a feakin dream job indeed. There's a reason you top the charts as one of the most desired places to work for 🤣

**Comment by etaham** [comment]
Any updates here? MSAL has been out for quite some time. Other tools, such as terraform, do not require the login call and are therefore more secure in this context.

**Comment by jiasli** [comment]
Nope. We'll make sure to update this thread when there is. In the meantime, might I recommend the **Subscribe** button?

![image](https://user-images.githubusercontent.com/4003950/155913610-6a788ba6-4bbe-4377-b1b4-efa48bc2e680.png)

That way you'll be notified of any updates to this thread, without needlessly pinging everyone on this thread. :)

**Comment by jiasli** [comment]
This feature aligns with the goal of Azure SDKs:

- https://github.com/Azure/azure-sdk-for-js/issues/5696
- https://github.com/Azure/azure-sdk-for-python/issues/13401
- https://github.com/Azure/azure-sdk/blob/main/_includes/tables/environment_variables.md

**Comment by skorobogatydmitry** [comment]
Any updates on this ?

**Comment by prem-mukherji** [comment]
Any updates on this?

**Comment by arunkumar611** [comment]
@jiasli - When the enhancement goes live, will it also support OIDC?  There are the variables that are using by other providers. 

"AZURE_OIDC_REQUEST_TOKEN": "[secret]",
 "AZURE_OIDC_REQUEST_URL": "https://api.x.com/oidc",
 "AZURE_OIDC_TOKEN": "[secret]",
 "AZURE_USE_OIDC": "true"

**Comment by jiasli** [comment]
> There are the variables that are used by other providers.

@arunkumar611, could you share more information on these "providers"? I am not able to find any reference to these environment variables on the internet: https://www.bing.com/search?q=AZURE_OIDC_REQUEST_URL , https://www.google.com/search?q=AZURE_OIDC_REQUEST_URL

**Comment by tspearconquest** [comment]
> > There are the variables that are used by other providers.
> 
> @arunkumar611, could you share more information on these "providers"? I am not able to find any reference to these environment variables on the internet: https://www.bing.com/search?q=AZURE_OIDC_REQUEST_URL , https://www.google.com/search?q=AZURE_OIDC_REQUEST_URL

I think that commenter may have meant `ARM_*` instead of `AZURE_*` as I do see these 4 on google:
```
"ARM_OIDC_REQUEST_TOKEN": "[secret]",
"ARM_OIDC_REQUEST_URL": "https://api.x.com/oidc",
"ARM_OIDC_TOKEN": "[secret]",
"ARM_USE_OIDC": "true"
```

**Comment by ringods** [comment]
@tspearconquest I thought so too, but proposed PR #27938 implements it with the `AZURE_` prefix.

**Comment by Speeddymon** [comment]
It's not the same variables at all. Those are strictly related to SP auth with secrets whereas this poster asked about OIDC vars but appears to have mistakenly used the AZURE_ prefix. That's all I was trying to say, but thank you. :)

**Comment by MattyKuzyk** [comment]
Still no updates? It's been 7 years. Most other modern CLIs I'm aware allow configuration via env vars (`aws` and `kubectl` to name a few).

Our repo depends heavily on `direnv` for making sure your configuration is correct depending on your directory. Azure CLI doesn't play nice with this at all and we had to design a wrapper script just to support env vars.
